### PR TITLE
acton 0.7.1 (new formula)

### PR DESCRIPTION
**Pull request by plajjan**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/acton.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  depends_on "util-linux"
+
+  on_macos do
+    depends_on "argp-standalone" => :build
+  end
+
+  on_linux do
+    depends_on "gcc"
+    depends_on "gmp"
+    depends_on "libbsd" => :build
```
Comment (carlocab) [inline Formula/acton.rb:22]:
```suggestion
    depends_on "libbsd" => :build
    depends_on "gcc"
    depends_on "gmp"
```

File: Formula/acton.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  depends_on "protobuf-c"
+  depends_on "utf8proc"
+  depends_on "util-linux"
+
+  on_macos do
+    depends_on "argp-standalone" => :build
+  end
+
+  on_linux do
+    depends_on "gcc"
```
Comment (carlocab) [inline Formula/acton.rb:20]: Why do we need this? Note that adding `depend_on "gcc"` is not enough to make `brew` use Homebrew GCC as the C/C++ compiler. (It'll be used for `gfortran`, though.)

File: Formula/acton.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    depends_on "gmp"
+    depends_on "libbsd" => :build
+  end
+
+  def install
+    ENV["BUILD_RELEASE"] = "1"
+    system "make"
+    bin.install "dist/bin/actonc"
+    bin.install "dist/bin/actondb"
+    prefix.install Dir["dist/*"]
```
Comment (carlocab) [inline Formula/acton.rb:30]: Is there no `make install`?

File: Formula/acton.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  test do
+    system "#{bin}/actonc", "--version"
+    (testpath/"hello.act").write <<~EOS
+      actor main(env):
+          print("Hello World!")
+          await async env.exit(0)
+    EOS
+    system "#{bin}/actonc", "--root", "main", "hello.act"
+    assert_equal "Hello World!\n", `./hello`
```
Comment (carlocab) [inline Formula/acton.rb:41]:
```suggestion
    assert_equal "Hello World!\n", shell_output("./hello")
```

**Review by carlocab — COMMENTED** [review]
Thanks for the PR, @plajjan. A few comments/questions.

File: Formula/acton.rb (hunk 2 above)
Comment (plajjan) [reply Formula/acton.rb:20]: Acton is a programming language. It uses C as an intermediate language and to compile the final binary executable we need a C compiler. Acton uses `cc`, so on Mac it is clang while on the typical Linux machine, it is gcc. On Mac we get clang by default so no need to declare a dependency while on Linux I think we must declare the dependency, no?

I've seen there is some statement for when using parts of the base system too, maybe that is appropriate to use here as well?

**Comment by plajjan** [comment]
@carlocab, since you've kindly reviewed this PR, I hope you don't mind me asking. How do I skip tests on Apple M1 silicon? Acton relies on Haskell stack, which isn't properly working on M1 just yet - I have it locally working by doing `brew install --HEAD stack` but AFAIK there is no way to represent a dependency on HEAD in a formula, so I can't modify the acton formula to get it to work. ISTM that we just have to wait for 1) stack to release a new version with fixes for M1 2) the homebrew recipe for stack to use that new version. Meanwhile, I would like to skip M1 / arm64, how can I do that?

File: Formula/acton.rb (hunk 3 above)
Comment (plajjan) [reply Formula/acton.rb:30]: Nope, there is not.

File: Formula/acton.rb (hunk 2 above)
Comment (carlocab) [reply Formula/acton.rb:20]: The situation is the same Linux: `brew` uses the host compiler (i.e. GCC) which is a requirement for using `brew` on Linux. See: https://docs.brew.sh/Homebrew-on-Linux#requirements

**Comment by carlocab** [comment]
> How do I skip tests on Apple M1 silicon?

You can skip some of them, but not all of them. In any case, even if we skip the builds on M1, we would still get this audit failure on Intel:
```
  acton:
    * GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)
  Error: 1 problem in 1 formula detected
```

**Comment by SMillerDev** [comment]
Hi, this seems like a nice software, but not notable enough for us to package in Homebrew for now. This is not a decision on the merits of the software, but we cannot distribute every single piece of software and have to draw a line on “notability” somewhere. Of course, we would be happy to reconsider in the future.

Please consider hosting it in a personal tap, which is very easy to do: https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

**Comment by plajjan** [comment]
@SMillerDev is there a more tangible requirement on notability or is it arbitrary? I seem to remember seeing 75 github stars in some homebrew document but can't find that any longer.

There is a tap for acton already (https://github.com/actonlang/homebrew-acton/) but since the tap bottle building is run on GitHub Acton runners, there is no support for bottling it up for the Apple M1. Being part of homebrew-core would mean it could use the excellent build infrastructure here to produce an M1 build, although it is still hinged upon a new stack release (as mentioned in an above comment).

**Comment by plajjan** [comment]
Oh, I see the comment from @carlocab now mentioning github stars etc. Nevermind that then ;)

**Comment by plajjan** [comment]
I know this is closed but I wonder if I could ask a question on form and function of the Formula. @carlocab / @SMillerDev any pointers appreciated.

The Acton compiler is written in Haskell. There is a stack (haskell build tool / package manager) config in the repository and normally we let stack install a local GHC in the project build directory. I have noticed in some other brew formulas for Haskell programs that stack is told to use the system GHC and a particular GHC version is listed as a `depends_on` in the Formula.

Are both options allowed? Is one preferred over the other?

I have tried to look for documentation but can't find anything specifically for stack installs of GHC.

**Comment by carlocab** [comment]
For a formula in your tap, you can handle this however you wish. Formulae in Homebrew/core should use a GHC that we ship. This avoids issues like `stack` not providing a GHC binary for M1. (At least, this was the case last I checked, which was a little while ago. Things may be different now.)

**Comment by plajjan** [comment]
@carlocab thank you very much, exactly what I was looking for.

Indeed, I did experience problems with stack not being able to install ghc on an M1. My way around that was to `brew install --HEAD stack` which brings in a new enough version of stack that is able to install GHC on M1. An install of head can't be expressed as depends_on in a formula, so I would have to wait for a new version of stack to be released etc. What you recommend works today, out of the box. Thank you!

**Pull request closed without merging**